tools/cat2table.ml

```diff
@@ -0,0 +1,952 @@
+  | _ -> false
+
+  (* A character that we know can't be part of a variable name *)
+  let sep = '$'
+
+  let to_alias s n =
+    s ^ String.make 1 sep ^ string_of_int n
+
+  let from_alias s =
+    match String.index_opt s sep with
+    | Some idx -> String.sub s 0 idx
+    | None -> s
+
+  let pp_expr_with_loc e = ASTUtils.exp2loc e |> Cache.extract
+  let pp_expr =
+    let rec pp_with_sep paran rec_paran sep es =
+      let str = String.concat sep (List.map (do_pp_expr rec_paran) es) in
+      if paran then Printf.sprintf "(%s)" str else str
+    and do_pp_expr paran = function
+    | Konst (_, (Empty _)) -> "empty"
+    | Konst (_, (Universe _)) -> "universe"
+    | Var (_, id) -> from_alias id
+    | Op1 (_, Comp, e) -> Printf.sprintf "~%s" (do_pp_expr true e)
+    | Op (_, Inter, es) -> pp_with_sep paran true " & " es
+    | Op (_, Union, es) -> pp_with_sep paran true " | " es
+    | Op (_, Diff, es) -> pp_with_sep paran true " \\ " es
+    | Op (_, Seq, es) -> pp_with_sep paran false "; " es
+    | Try (_, e1, e2) ->
+      Printf.sprintf "try %s with %s" (do_pp_expr false e1) (do_pp_expr false e2)
+    | If (_, _, e1, e2) ->
+      Printf.sprintf "if _some_cond then %s else %s" (do_pp_expr false e1) (do_pp_expr false e2)
+    | e -> pp_expr_with_loc e in
+    do_pp_expr false
+
+  (** Comparison function on expressions - assumes that sub-expression in op
+    case are sorted based on itself *)
+  let compare_expr e1 e2 =
+    let is_commutative = function
+    | Union | Inter | Add -> true
+    | _ -> false in
+
+    let do_compare_pairs (a1, b1) (a2, b2) cmp1 cmp2 =
+      let c = cmp1 a1 a2 in
+      if c != 0 then c
+      else cmp2 b1 b2 in
+
+    let rec do_compare_expr e1 e2 = match e1, e2 with
+    | e1, e2 when is_empty e1 && is_empty e2 -> 0
+    | Konst (_, (Universe _)), Konst (_, (Universe _)) -> 0
+    | e, _ when is_empty e -> -1
+    | _, e when is_empty e -> 1
+    | Konst (_, (Universe _)), _ -> -1
+    | _, Konst (_, (Universe _)) -> 1
+    | Var (_, id1), Var (_, id2) -> String.compare id1 id2
+    | Var _, _ -> -1
+    | _, Var _ -> 1
+    | Op1 (_, op1, e1), Op1 (_, op2, e2) ->
+      do_compare_pairs (op1, e1) (op2, e2) compare do_compare_expr
+    | Op1 _, _ -> -1
+    | _, Op1 _ -> 1
+    | Op (_, op1, es1), Op (_, op2, es2) ->
+      do_compare_pairs (op1, es1) (op2, es2) compare do_compare_expr_list
+    | Op _, _ -> -1
+    | _, Op _ -> 1
+    | App (_, e11, e12), App (_, e21, e22) ->
+      do_compare_expr_pairs (e11, e12) (e21, e22)
+    | App _, _ -> -1
+    | _, App _ -> 1
+    | Try (_, e11, e12), Try (_, e21, e22) ->
+      do_compare_expr_pairs (e11, e12) (e21, e22)
+    | Try _, _ -> -1
+    | _, Try _ -> 1
+    | If (_, c1, e11, e12), If (_, c2, e21, e22) ->
+      do_compare_pairs (c1, (e11, e12)) (c2, (e21, e22))
+        do_compare_cond
+        do_compare_expr_pairs
+    | If _, _ -> -1
+    | _, If _ -> 1
+    | _, _ -> 0 (* We don't care about the rest, since they are not effects *)
```

Comment:
You can fall back to the `Stdlib.compare`. 

In fact many of `do_compare_*` function you introduce here probably can use the default `Stdlib.compare`, if you do not care a particular order but just need a total order over data structure.